libs/wingsdk/src/target-tf-azure/function.ts

```diff
@@ -0,0 +1,174 @@
+import * as fs from "fs";
+import { resolve } from "path";
+import { LinuxFunctionApp } from "@cdktf/provider-azurerm/lib/linux-function-app";
+import { ResourceGroup } from "@cdktf/provider-azurerm/lib/resource-group";
+import { RoleAssignment } from "@cdktf/provider-azurerm/lib/role-assignment";
+import { ServicePlan } from "@cdktf/provider-azurerm/lib/service-plan";
+import { StorageAccount } from "@cdktf/provider-azurerm/lib/storage-account";
+import { StorageBlob } from "@cdktf/provider-azurerm/lib/storage-blob";
+import { AssetType, TerraformAsset } from "cdktf";
+import { Construct } from "constructs";
+import { App } from "./app";
+import { Bucket } from "./bucket";
+import * as cloud from "../cloud";
+import * as core from "../core";
+import { NameOptions, ResourceNames } from "../utils/resource-names";
+
+/**
+ * Function names are limited to 32 characters.
+ * You can use alphanumeric characters.
+ * https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules#microsoftweb
+ */
+const FUNCTION_NAME_OPTS: NameOptions = {
+  maxLen: 32,
+  disallowedRegex: /[^a-z0-9]+/g,
+};
+
+/**
+ * Azure implementation of `cloud.Function`.
+ *
+ * @inflight `@winglang/wingsdk.cloud.IFunctionClient`
+ */
+export class Function extends cloud.FunctionBase {
+  private readonly function: LinuxFunctionApp;
+  private readonly app: App;
+  private readonly servicePlan: ServicePlan;
+  private readonly storageAccount: StorageAccount;
+  private readonly resourceGroup: ResourceGroup;
+  private permissions?: Map<string, string[]>;
+
+  constructor(
+    scope: Construct,
+    id: string,
+    inflight: cloud.IFunctionHandler,
+    props: cloud.FunctionProps
+  ) {
+    super(scope, id, inflight, props);
+
+    this.app = App.of(this);
+    this.storageAccount = this.app.storageAccount;
+    this.resourceGroup = this.app.resourceGroup;
+
+    const functionName = ResourceNames.generateName(this, FUNCTION_NAME_OPTS);
+
+    // Create service plan for function using dynamic sku
+    // https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/service_plan#sku_name
+    this.servicePlan = new ServicePlan(this, "ServicePlan", {
+      name: `serviceplan${functionName}`,
+      resourceGroupName: this.resourceGroup.name,
+      location: this.resourceGroup.location,
+      osType: "Linux",
+      skuName: "Y1",
```

Comment:
Also as a best practice it would be good to define these as constants at the top of the file (helps encourage contributors to use the constant if it ever needs to be reused in a second or third place instead of copying the string)